### PR TITLE
stm32mp1: introduce platform clock driver

### PR DESCRIPTION
**Pull request by etienne-lms**
Driver API uses the clock IDs published in the DT bindings (as per later Linux kernel v4.20) to identify the clocks.

Note configuration of clock parenthood and related divisors as well as configuration of the PLLs are under to responsibility of an earlier boot stage. Such configuration directives found by OP-TEE core in the DT are ignored. DT is used to set the SoC clocks interface secure state and the frequencies of the platform source clocks.

Driver API allows to gate (enable/disable) a clock, get a clock enable state and its frequency.

### CODE REVIEW

File: core/arch/arm/plat-stm32mp1/shared_resources.c [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+/*
+ * Copyright (c) 2017-2018, STMicroelectronics
+ */
+
+#include <drivers/stm32mp1_clk.h>
+#include <stm32_util.h>
+
+unsigned long stm32_clock_get_rate(unsigned long id)
+{
+	return stm32mp1_clk_get_rate(id);
```
Comment (jforissier) [inline core/arch/arm/plat-stm32mp1/shared_resources.c]: Why have two names and this empty wrapper? You could directly use the name `stm32_clock_get_rate()` in the driver I suppose?

Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/shared_resources.c]: You're right regarding `_get_rate()` and `_is_clock_enabled()` but not regarding `_clock_enable()` and `_clock_disable()`. To 2 later will (soon) integrate a specific sequence related to shared clock support.
Looks like I should simply drop the indirection for the 2 former. Thanks for the feedback. I'll push a fixup.

Comment (jforissier) [reply core/arch/arm/plat-stm32mp1/shared_resources.c]: Sorry, I don't get why the two others are different. I think you should add the additional layer only when it becomes necessary.

Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/shared_resources.c]: :) 
let's say this allows to introduce the **shared_resources.c** and it's handy regarding the other P-Rs I want to submit (i.e https://github.com/etienne-lms/optee_os/pull/13).

Comment (jforissier) [reply core/arch/arm/plat-stm32mp1/shared_resources.c]: Maybe, but it does not change the fact that I'd rather have logical and minimal commits merged into master. Not "stuff that I will probably need very soon, promised, you will understand when you see it!" ;-)
Renaming and moving some code **when it's needed** is perfectly fine. Why create a file called `shared_ressources.c` when nothing is shared yet? It is confusing. Create it when you actually have something to share.

Comment (etienne-lms) [reply core/arch/arm/plat-stm32mp1/shared_resources.c]: Ok, I get you point view. Fixing...

**Comment by etienne-lms** [comment]
hmmm, this needs some rebasing since https://github.com/OP-TEE/optee_os/pull/2748 is merged.
I'll wait you are fine with the content and will rebase/squash by then.

**Comment by jforissier** [comment]
> I'll wait you are fine with the content and will rebase/squash by then.

Please do, this LGTM now. Thanks.

**Comment by etienne-lms** [comment]
ok, fair. thanks. and sorry for bothering you with my bad habits.

**Comment by etienne-lms** [comment]
rebased and squashed + fixed from comment in 2nd commit.
@jforissier I applied your tag on the 1st and 3rd commits. No change since you acked them.

**Comment by jforissier** [comment]
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by etienne-lms** [comment]
Tag applied + added credits for the driver contributors on 2nd commit.

Note: [minor late changes](https://github.com/OP-TEE/optee_os/files/2839722/stm32mp1_clk.c-changes.txt) (comment removal) in the last commit comment, yet I kept your tag. Tell me if that's an issue.

**Pull request merged**